### PR TITLE
DS-4210 Deposit on collection without workflow seems to fail (405 but succeed)	

### PR DESCRIPTION
**Pull request by abollini**
I have created a test to demonstrate the issue https://github.com/DSpace/DSpace/commit/8556441f43d52bf7834776a6ed6cdefa7d029c54

and the subsequent fix that looks quite obvious to me but the automatic test is still failing
https://travis-ci.org/4Science/DSpace/builds/527325143#L5750

Testing the fix manually via the angular UI everything looks good so I guess there is some oddity in the test env cleanup process

### CODE REVIEW

**Comment by cwilper** [comment]
@abollini I checked out the latest code in this branch and attempted a build. I got several errors, including the `WorkflowItemRestRepositoryIT ` one you referenced above. An error does occur in test cleanup. That type of error ("Batch update returned unexpected row count") can happen when the underlying db connection is in a rolled back state and therefore cannot commit.

The db connection was probably auto-rolled back due to this constraint violation:

```
Caused by: org.h2.jdbc.JdbcSQLException: Unique index or primary key violation: "CONSTRAINT_CE8_INDEX_6 ON PUBLIC.EPERSON(EMAIL) VALUES ('submitter@example.com', 308)"; SQL statement:
update PUBLIC.eperson set can_log_in=?, digest_algorithm=?, email=?, last_active=?, netid=?, password=?, require_certificate=?, salt=?, self_registered=?, session_salt=? where uuid=? [23505-187]
```

This violation tells me that at some point, either in the life of this transaction or earlier, an eperson was created with the same email, but it wasn't cleaned up as expected.

_Where is the code that cleans up the test epersons?_

I did find the following provocative bit in `AbstractIntegrationTestWithDatabase#destroy`:

```
            // NOTE: we explicitly do NOT destroy our default eperson & admin as they
            // are cached and reused for all tests. This speeds up all tests.
```

So it seems that subclasses would be responsible for cleaning up after any *other* epersons they create.

**Comment by cwilper** [comment]
@abollini I wanted to let you know I also just ran `depositWorkspaceItemWithoutWorkflowTest` in isolation and DID NOT see the same constraint violation I quoted above. I also learned where EPersons (and other DSOs) are cleaned up by the builders. I am still consistently seeing the hibernate/batch error when it's trying to clean up the test collection.

```
java.lang.RuntimeException: javax.persistence.OptimisticLockException: Batch update returned unexpected row count from update [3]; actual row count: 0; expected: 1

	at org.dspace.app.rest.test.AbstractIntegrationTestWithDatabase.destroy(AbstractIntegrationTestWithDatabase.java:193)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:252)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:94)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: javax.persistence.OptimisticLockException: Batch update returned unexpected row count from update [3]; actual row count: 0; expected: 1
	at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:212)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:86)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:155)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:162)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1434)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1414)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.hibernate.context.internal.ThreadLocalSessionContext$TransactionProtectionWrapper.invoke(ThreadLocalSessionContext.java:355)
	at org.dspace.core.HibernateDBConnection.commit(HibernateDBConnection.java:94)
	at org.dspace.core.Context.commit(Context.java:421)
	at org.dspace.core.Context.complete(Context.java:373)
	at org.dspace.app.rest.builder.AbstractDSpaceObjectBuilder.delete(AbstractDSpaceObjectBuilder.java:230)
	at org.dspace.app.rest.builder.CollectionBuilder.cleanup(CollectionBuilder.java:117)
	at org.dspace.app.rest.builder.util.AbstractBuilderCleanupUtil.cleanupBuilders(AbstractBuilderCleanupUtil.java:86)
	at org.dspace.app.rest.builder.AbstractBuilder.cleanupObjects(AbstractBuilder.java:170)
	at org.dspace.app.rest.test.AbstractIntegrationTestWithDatabase.destroy(AbstractIntegrationTestWithDatabase.java:177)
```

I did notice in the code, that when the test collection cleanup occurs, it ends up deleting a lot of things transitively. It seems that maybe one of those things it's trying to delete(possibly the collection's core resource policies?) is stale in the hibernate session. I had trouble narrowing in on exactly what was stale, but when turning on sql logging in hibernate, I saw the last updates it was trying where DELETEs from the resourcepolicy table.

**Comment by abollini** [comment]
@cwilper thanks for your suggestions it point me in the right direction for debugging the issue. At the end the issue was related to a different problem (the cleanup process for workspaceitem and workflowitem that when "moved around" left reference that also need to be removed). The PR is now ready to be reviewed

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java [hunk 1]
```diff
@@ -101,22 +106,48 @@ private void deleteWsi(WorkspaceItem dso) throws Exception {
...
+            c.complete();
+        }
+
+        indexingService.commit();
+    }
 
     @Override
     public void cleanup() throws Exception {
         if (workspaceItem != null) {
+            System.out.println("remove workspaceitem");
```
Comment (tdonohue) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java]: We should remove this `System.out.println` ;)

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java [hunk 2]
```diff
@@ -101,22 +106,48 @@ private void deleteWsi(WorkspaceItem dso) throws Exception {
...
             deleteWsi(workspaceItem);
         }
         if (workflowItem != null) {
+            System.out.println("remove workflowitem");
```
Comment (tdonohue) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java]: Again, not needed anymore.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java [hunk 3]
```diff
@@ -101,22 +106,48 @@ private void deleteWsi(WorkspaceItem dso) throws Exception {
...
             delete(workflowItem);
         }
+        if (item != null) {
+            System.out.println("remove item!!!");
```
Comment (tdonohue) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java]: Again,  please remove this println

**Review by tdonohue — APPROVED** [review]
👍 Overall, the code here looks good. I haven't tested it, but the ITs look good. I have very minor inline comments about removing three `println` statements which are no longer needed.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java (hunk 1 above)
Comment (mwoodiupui) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkflowItemBuilder.java]: If we want to keep some notice of what the code is doing, maybe change these to log.debug().

**Comment by abollini** [comment]
thanks for the review @tdonohue  @mwoodiupui 
I removed the sysout spot by tim as they were only needed during debugging, it doesn't make sense to turn them in log debug line so far (or they should be more descriptive)

**Review by benbosman — APPROVED** [review]
The code looks correct to me, and I didn't see any errors when testing the functionality

**Comment by tdonohue** [comment]
This is at +2, so I'm merging it.  Thanks @abollini  (and @cwilper for the hints to help get this solved)!

**Pull request merged**